Plugins: Display live chat support only for annual plans

epeicher opened the pull request:
The live chat support message when purchasing a plugin will only be
displayed for annual plans regardless of the billing period of the
plugin.

#### Changes proposed in this Pull Request

* Displays `Live chat support` message only for annual plans
* For monthly plans is displayed the message `Unlimited email support`

|Before | After|
|-------|------|
|![CleanShot 2022-02-22 at 18 06 04@2x](https://user-images.githubusercontent.com/3519124/155182231-1075cf1d-261f-4466-9408-5474937d2ae0.png)|![CleanShot 2022-02-22 at 18 14 24@2x](https://user-images.githubusercontent.com/3519124/155183559-990d74b5-d988-44f4-9b69-10aaa622f272.png)|
|![CleanShot 2022-02-22 at 18 07 17@2x](https://user-images.githubusercontent.com/3519124/155182414-bd8f7565-f082-4872-bc64-b02445f706d2.png)|![CleanShot 2022-02-22 at 18 16 35@2x](https://user-images.githubusercontent.com/3519124/155183933-1ac5a036-c42b-43ae-9dab-989d1b6318ef.png)|
|![CleanShot 2022-02-22 at 18 11 04@2x](https://user-images.githubusercontent.com/3519124/155183047-e35c83b1-4c31-4cb4-8d06-ef6ca0aa8038.png)|![CleanShot 2022-02-22 at 18 17 47@2x](https://user-images.githubusercontent.com/3519124/155184165-69292d5c-0050-4b87-a25c-77f1f0d6f1f1.png)|
|![CleanShot 2022-02-22 at 18 12 09@2x](https://user-images.githubusercontent.com/3519124/155183183-c2fc5913-1ab2-484e-8e83-e5945f01c009.png)|![CleanShot 2022-02-22 at 18 18 36@2x](https://user-images.githubusercontent.com/3519124/155184294-b649c135-3d69-4a2a-be72-959ab9445d00.png)|

#### Testing instructions

##### Scenario 1
* In a site with either Annual Business or eCommerce plan which is paid **Monthly**
* Add a paid plugin
* Select Monthly billing for the plugin
  * The details of the Purchase and the Support details on the right should include `Unlimited email support`
* Select Annually billing for the plugin
  * The details of the Purchase and the Support details on the right should include `Unlimited email support`

##### Scenario 2
* In a site with either Annual Business or eCommerce plan which is paid **Annually**
* Add a paid plugin
* Select Monthly billing for the plugin
  * The details of the Purchase and the Support details on the right should include `Live chat support 24x7` 
* Select Annually billing for the plugin
  * The details of the Purchase and the Support details on the right should include `Live chat support 24x7`

##### Screenshots
|Monthly Plan|Annual Plan|
|---|---|
|![CleanShot 2022-02-23 at 09 12 51@2x](https://user-images.githubusercontent.com/3519124/155282086-faed9ae1-be51-49d6-b2d8-80663909a7bf.png)|![CleanShot 2022-02-23 at 09 10 56@2x](https://user-images.githubusercontent.com/3519124/155281868-1843d5de-0393-4f06-bba1-6f121f586ca0.png)|

Fixes #61365

epeicher committed:
Display live chat support only for annual plans

The live chat support message when purchasing a plugin will only be
displayed for annual plans regardless of the billing period of the
plugin.

Fixes #61365

cpapazoglou commented on client/my-sites/plugins/plugin-details-CTA/usps.tsx:
```diff
@@ -48,6 +50,12 @@ const USPS: React.FC< Props > = ( {
 
+	const selectedSiteId = useSelector( getSelectedSiteId );
+	const currentPlan = useSelector( ( state ) =>
+		selectedSiteId ? getCurrentPlan( state, selectedSiteId ) : undefined
```
```suggestion
		selectedSiteId && getCurrentPlan( state, selectedSiteId )
```

would that work?

cpapazoglou commented on client/my-sites/plugins/plugin-details-sidebar/index.jsx:
```diff
@@ -19,8 +20,12 @@ const PluginDetailsSidebar = ( {
+
+	const selectedSiteId = useSelector( getSelectedSiteId );
+	const currentPlan = useSelector( ( state ) =>
+		selectedSiteId ? getCurrentPlan( state, selectedSiteId ) : undefined
```
```suggestion
		selectedSiteId && getCurrentPlan( state, selectedSiteId )
```

would that work?

cpapazoglou reviewed (changes requested):
The scenarios work as expected but there is one more scenario that we need to take care of:

On site plans that need to be upgraded to business (free, personal, premium) then we will need to check against the billing plan interval if the the current plan is monthly cause that is used to add the correct billing interval of a business to cart.

- If current plan is annual, then users cannot upgrade to monthly busines
- if current plan is monthly, then users can upgrade to monthly business. In that case we shouldn't get Live chat support cause a monthly business plan will get added

![CleanShot 2022-02-24 at 13 20 52@2x](https://user-images.githubusercontent.com/12430020/155515091-e984c55b-07d3-4336-8386-f68ae6eb6ade.png)

epeicher committed:
Apply suggestions from code review

Co-authored-by: Harris Papazoglou <hrrsppzgl@gmail.com>

epeicher committed:
Fix style after applying code review suggestions

epeicher commented:
Additional changes to cope with upgrades are described in the screenshots below. 

@cpapazoglou could you please confirm if those scenarios work as expected?

|Plan Period|Plugin Monthly | Plugin Annually|
|-------|-------|------|
|Annual|![CleanShot 2022-02-25 at 11 45 27@2x](https://user-images.githubusercontent.com/3519124/155703015-02bcbd88-ef2d-4bd9-b224-0c651d73293c.png)|![CleanShot 2022-02-25 at 11 46 24@2x](https://user-images.githubusercontent.com/3519124/155703082-67f93a8b-4e0e-4421-8e31-b9a6561b6927.png)|
|Monthly|![CleanShot 2022-02-25 at 11 54 42@2x](https://user-images.githubusercontent.com/3519124/155703383-270d24d9-bdc1-4bfe-8f56-817512d45ef3.png)|![CleanShot 2022-02-25 at 11 55 42@2x](https://user-images.githubusercontent.com/3519124/155703474-5d3a26c1-c260-49c7-97bf-a33d57fd105c.png)|

⚠️ **Please note** that in the scenario with current plan as Annual, even though the Monthly plugin option is displayed (top-left screenshot above), when clicking on `Purchase and activate` button, only the Annual plan is displayed:
<img width="1126" alt="CleanShot 2022-02-25 at 11 59 37@2x" src="https://user-images.githubusercontent.com/3519124/155704220-f0be5124-cfc8-4b7d-b481-00b6b85e9371.png">

epeicher committed:
Add an additional condition to cope with upgrades

When the upgrade information is displayed, showing the live chat support
will depend on the billing period of the plugin as that will subsequently
derive the period of the plan to be upgraded.

gcsecsey reviewed (changes requested):
I've tested the scenarios described, and found some which aren't working as expected yet:

When testing **Scenario 1** with a monthly billed business plan, the CTA and sidebar USPs show different values:
![image](https://user-images.githubusercontent.com/11555574/156155567-a5593eb1-2f22-4879-87a3-58048fa22dea.png)

Similarly, when testing with a **Free plan**:
![image](https://user-images.githubusercontent.com/11555574/156155754-c03eaa3a-a9fd-4ad9-b1e5-054f34b6adaf.png)

This is most likely because the sidebar doesn't yet take into account the `shouldUpgrade` flag.

gcsecsey committed:
proposal: use the same logic in sidebar and under CTA button to display support USP

gcsecsey commented:
I've pushed some proposed changes, to use the same logic in both places. Moreover, I think we could move this logic into a selector (eg. under `state/marketplace/selectors`), which would make it more easy and robust to use this logic.

@epeicher what do you think? Let me know if I can help! 🙂

epeicher commented:
> I've pushed some proposed changes, to use the same logic in both places. Moreover, I think we could move this logic into a selector (eg. under `state/marketplace/selectors`), which would make it more easy and robust to use this logic.
> 
> @epeicher what do you think? Let me know if I can help! 🙂

Thank you very much @gcsecsey! Also, the idea of a common selector sounds great, I will work on it

epeicher committed:
Extract common logic to a selector

That will make the logic reusable and robust as the same logic will be
shared by the consumers

gcsecsey reviewed (approved):
I've tested all scnearios again, LGTM!

I only have one small proposal, I think the `isAnnualPeriod` is no longer needed, as `isAnnualPlan` has the same value.

gcsecsey commented on client/my-sites/plugins/plugin-details-CTA/usps.tsx:
```diff
@@ -46,8 +47,9 @@ const USPS: React.FC< Props > = ( {
 	const translate = useTranslate();
 
-	const isAnnualPeriod = billingPeriod === IntervalLength.ANNUALLY;
+	const isAnnualPlan = useSelector( isAnnualPlanOrUpgradeableAnnualPeriod );
 
+	const isAnnualPeriod = billingPeriod === IntervalLength.ANNUALLY;
 	const planDisplayCost = useSelector( ( state ) =>
 		getProductDisplayCost( state, isAnnualPeriod ? PLAN_BUSINESS : PLAN_BUSINESS_MONTHLY )
```
```suggestion
	const isAnnualPlan = useSelector( isAnnualPlanOrUpgradeableAnnualPeriod );

	const planDisplayCost = useSelector( ( state ) =>
		getProductDisplayCost( state, isAnnualPlan ? PLAN_BUSINESS : PLAN_BUSINESS_MONTHLY )
```

epeicher replied:
We would be changing previous logic in that case because `isAnnualPeriod` contains the billing period selected when purchasing a _plugin_ while the new `isAnnualPlan` contains logic related to the _plan_ selected by the user. 

In other words, there is a scenario where `isAnnualPeriod` could be `true` while `isAnnualPlan` could be `false`. For example, if the user has a _Business Monthly_ plan but selects _Annual_ billing period when purchasing the plugin.

gcsecsey replied:
That's a good point! My thinking was that the new selector kind of excludes this scenario, because we return either `isAnnualPeriod` or `isAnnualPlan` based on the value of `shouldUpgrade`.

As the only place it's then used is also guarded with `shouldUpgrade`, I'm good with going forward with either version! 🎉 🚢

epeicher replied:
That is totally correct! Although, in this scenario, I would prefer to leave the current version for `planDisplayCost` because it reflects a more specific condition. So if you are happy with the current version, I will go ahead and merge the issue.

Thanks for your comments!
